Gate of the fold Fixes and minor clean up

Phil-M-H opened the pull request:
1. Removing a PortalLocation did not function properly, resulting portals teleporting to already destroyed portals. Fixed with line 175 of RitualPortal

2. The 1st portal added into a world would not function properly, So when you created your first 2 portals and tried to initialize one, then the other, they wouldn't work, and you would instead have to use the activation crystal on the 1st one a second time. 

Fixed making the LocationsHandler properly static. I was unsure of what the getLocationsHandler() method was supposed to do - I'm assuming to verify that the various variables had already been loaded correctly from the file, so I tried to keep that functionality with the verifyIsInitialized() and the initialized boolean. 

3. Having multiple gates was extremely inconsistent, with some of the previous gates still having their portal tiles but not being functional. I changed the functionality so that all gates teleport to the most recently activated gate with the same key. So if you have 5 gates, they will all teleport to the most recently activated gate. The most recently activated gate will teleport to the 2nd most recently activated gate.

Phil-M-H committed:
Gate of the fold complete

Phil-M-H committed:
Remove unused imports

keraldi commented:
#1512

TehNut commented on src/main/java/WayofTime/bloodmagic/block/BlockDimensionalPortal.java:
```diff
@@ -76,27 +76,24 @@ public void onEntityCollision(World world, BlockPos pos, IBlockState blockState,
+                    IMasterRitualStone masterRitualStone = (IMasterRitualStone) world.getTileEntity(tile.getMasterStonePos());
+                    PortalLocation portal;
+                    int index = linkedLocations.size() - 1; //index of most recent PortalLocation
+                    if (linkedLocations.get(index).equals(new PortalLocation(masterRitualStone.getBlockPos().up(), world.provider.getDimension()))) {
```
https://github.com/WayofTime/BloodMagic/blob/4c76d42ef4a0cd37e70df38acde31e8d0542dfff/src/main/java/WayofTime/bloodmagic/block/BlockDimensionalPortal.java#L87-L91

-->

```java
if (linkedLocations.get(index).equals(new PortalLocation(masterRitualStone.getBlockPos().up(), world.provider.getDimension())))
    index--; // if most recent PortalLocation = this, get the 2nd most recent
portal = linkedLocations.get(index);
```

TehNut commented on src/main/java/WayofTime/bloodmagic/block/BlockDimensionalPortal.java:
```diff
@@ -76,27 +76,24 @@ public void onEntityCollision(World world, BlockPos pos, IBlockState blockState,
+            ArrayList<PortalLocation> linkedLocations = LocationsHandler.getLinkedLocations(tile.portalID);
+
+            if (linkedLocations != null && linkedLocations.size() > 1) {
+                if (world.getTileEntity(tile.getMasterStonePos()) != null && world.getTileEntity(tile.getMasterStonePos()) instanceof IMasterRitualStone) {
```
`instanceof` implies nonnull. The tile obtained here should also be stored to avoid unnecessary lookups

TehNut commented on src/main/java/WayofTime/bloodmagic/block/BlockDimensionalPortal.java:
```diff
@@ -76,27 +76,24 @@ public void onEntityCollision(World world, BlockPos pos, IBlockState blockState,
-                                TeleportQueue.getInstance().addITeleport(new Teleports.TeleportToDim(portal.getX(), portal.getY(), portal.getZ(), entity, masterRitualStone.getOwner(), world, portal.getDimension(), false));
-                            }
-                        }
+            LocationsHandler.verifyIsInitialized();
```
Probably better to call this once when the server loads rather than in 3 different places.

Phil-M-H replied:
How should I go about that, an event handler for serverStarting, or put it in GenericHandler's  onWorldLoad?

Phil-M-H committed:
Removed the null check in BlockDimensionalPortal
Neither ServerStartingEvent or ServerStartedEvent, were able to
initialized the LocationsHandler correctly, so it was placed in onWorldLoad

Phil-M-H committed:
Removed unused import and method.

TehNut commented on src/main/java/WayofTime/bloodmagic/ritual/types/RitualPortal.java:
```diff
@@ -172,7 +170,7 @@ public void stopRitual(IMasterRitualStone masterRitualStone, BreakType breakType
         }
 
-        portalRitualTag.removeTag(PORTAL_ID_TAG);
+        portalRitualTag.removeTag(portalRitualTag.getString(PORTAL_ID_TAG));
```
I keep looking at this and I can't help but feel it's wrong. Does `removeTag` not take the key?

Phil-M-H replied:
It does not. That was what probably was the most annoying to track down, because it just wouldn't remove previous portal locations.

A-UNDERSCORE-D commented:
you have an extra semicolon at https://github.com/WayofTime/BloodMagic/pull/1553/files#diff-81ae3006fe3f9b05ca3f89855c43a9ceR471

Phil-M-H committed:
Removed extra semi-colon.

keraldi commented:
Still broken.
With multiple portals set up, they all direct to the one created last.

Phil-M-H commented:
> Still broken.
> With multiple portals set up, they all direct to the one created last.

I designed it so that portals with identical keys would all teleport to the last portal with that key created -- and the last portal with that key would direct to the second to last.

It is less buggy now, but it still needs additions like factoring in all sides that touch the portal, and keeping track of the player's direction with respect to the portal. Probably another two weeks or more until I will have the time, unfortunately.

keraldi committed:
Merge branch '1.12' into gate-of-the-fold
